Check for invalid typeids

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

<!-- Describe your changes in detail. -->
On initialization, check for invalid type ids.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
When the user initializes a system with type ids that are invalid, various portions of HOOMD may access out of bounds memory and run invalid simulations or cause a segmentation fault.

This PR provides users with an understandable error message on initialization. Note that local snapshots provide direct access to underlying data, so it is still possible for users to set invalid type ids.

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #838 

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
This PR adds a unit test.

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Changed:

- Raise an exception when initializing systems with invalid particle type ids.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

bdice commented on hoomd/ParticleData.cc:
```diff
@@ -1188,6 +1194,24 @@ void ParticleData::initializeFromSnapshot(const SnapshotParticleData<Real>& snap
+        }
+    #endif
+
+    if (max_typeid >= m_type_mapping.size())
```
Is it valid to have a system with zero particles and zero types? If so, this would throw an error because `0 >= 0`. That might be undesired.

joaander replied:
It is valid to have a gsd file with 0 particles in a frame. I would need to check and see how HOOMD behaves with 0 particles in the whole system. My guess is most of the code will behave fine - it does need to support 0 particles on an individual MPI rank.

I would guess that most code in HOOMD assumes there is at least one particle type. Do you think it would be valid to run a simulation with 0 particle types? There would be no way to specify pair interaction potentials. I guess you could have 0 particle types but 1 bond type and model gaussian chains... I would have to test and see whether those code paths allow that and whether those that require at least 1 type raise appropriate errors.

mphoward replied:
I regularly run HOOMD with no HOOMD particles and only MPCD particles. I removed a few checks on number of particles when I implemented MPCD to support this, but in general, everything seemed to work fine because ranks could already have zero particles.

I could see it being reasonable to have 0 particle types if and only if you have no HOOMD particles in a system. I don't think you can have a bond without particles, though, since bonds are defined between two particles; these would need to have a type.

joaander replied:
> I could see it being reasonable to have 0 particle types if and only if you have no HOOMD particles in a system. I don't think you can have a bond without particles, though, since bonds are defined between two particles; these would need to have a type.

This makes sense. I will update the check accordingly.

b-butler commented on hoomd/pytest/test_snapshot.py:
```diff
@@ -309,3 +309,16 @@ def test_from_gsd_snapshot_populated(s, device):
     assert_equivalent_snapshots(gsd_snap, hoomd_snap)
+
+
+def test_invalid_particle_types(simulation_factory, lattice_snapshot_factory):
```
I think we should emphasis the distinction between `types` and `typeids`.
```suggestion
def test_invalid_particle_typeids(simulation_factory, lattice_snapshot_factory):
```

b-butler commented on hoomd/pytest/test_snapshot.py:
```diff
@@ -309,3 +309,16 @@ def test_from_gsd_snapshot_populated(s, device):
+
+
+def test_invalid_particle_types(simulation_factory, lattice_snapshot_factory):
+    """Test that using invalid particle types raises an error."""
```
```suggestion
    """Test that using invalid particle typeids raises an error."""
```

b-butler reviewed (approved):
How does a person know what are valid values for `typeid`? I don't see the answer in the API documentation? The implementation seems to work.

joaander commented:
@cbkerr feel free to review if you like, but I did add you just as an example so don't feel obligated.

joaander commented on hoomd/ParticleData.cc:
```diff
@@ -987,6 +988,9 @@ void ParticleData::initializeFromSnapshot(const SnapshotParticleData<Real>& snap
                 N_proc[rank]++;
+
+                // determine max typeid on root rank
+                max_typeid = std::max(max_typeid, snapshot.type[snap_idx]);
```
Thanks. This would be a better name, but we can solve that in a larger code refactoring PR.

joaander commented:
I need to add a corresponding check for `BondedGroup` typeids.

joaander commented:
I was mistaken, `BondedGroupData.cc` did already have a check for invalid typeids. I cleaned up the error messages to include the full message in the exception reason.

I have addressed all comments, this should be good to merge.

joaander commented:
/azp run validate
